### ISSUE TITLE
Suppport comments in WebExtension JSON's

### ISSUE THREAD

**Issue by rugk** [WeblateOrg/weblate#2043]
So, as turns out you do not actually support the WebExtensions JSON's (https://github.com/WeblateOrg/weblate/issues/1967) completely…
**Edit:** Current doc: https://docs.weblate.org/en/latest/formats.html#webex

You do not support comments.

However – in contrast to usual JSON files – they are actually supported in WebExtensions, at least in Chrome/ium and Firefox. 

* https://discourse.mozilla.org/t/manifest-json-should-mention-it-allows-comments/27529
* it's even in the spec: https://browserext.github.io/browserext/#manifest-json
* https://developer.mozilla.org/en-US/Add-ons/WebExtensions/manifest.json

**Comment by nijel** [comment]
This has to be addressed in the translate toolkit, library we use for parsing files.

Anyway I find this silly, I always thought it *is* JSON:

> Disregarding comments, the manifest.json file format for browser extensions is expected to be fully JSON compliant.

**Comment by nijel** [comment]
Anyway the docs is not really clear in this - the links you've posted are about `manifest.json`, not about `messages.json`, where no such exception is defined.

**Comment by rugk** [comment]
Indeed, maybe the spec does not specify it, but well… browsers support it.

**Comment by rugk** [comment]
An idea for an kinda ugly hack to make it work:

When importing:
1. Convert comments to some special translation ID you now so it can be represented in JSON, e.g. "// comment 1 2 3" to "{translation_NBGIJWGZI}" (generally WebExtension translation files have a quite strict format, so it is easy  to distinguish these)
2. Exclude them from translations and only use them behind-the-scenes.
3. When exporting just do the same vice versa.

---

Also BTW, do you actually preserve new lines in the translation files? Because I actually like to group translations. Combined with comments this makes much sense, IMHO.

**Comment by nijel** [comment]
Currently no formatting is preserved in the JSON, it's always fully generated.

**Comment by rugk** [comment]
Okay, created new issue for that: https://github.com/WeblateOrg/weblate/issues/2066

**Comment by rugk** [comment]
BTW one way to just remove all comments is to pass it through JSMin: https://plus.google.com/+DouglasCrockfordEsq/posts/RK8qyGVaGSr

**Comment by rugk** [comment]
Also note for the display in Weblate, it would be good, if it would use them (the coments) as section/group headers, just like this application does:
https://github.com/Lusito/web-ext-translator/blob/master/README.md

**Comment by nijel** [comment]
Yes, but it would discard them on saving... These days there might be some json5 library suitable for this, but it needs some research to find which of the dozen Python modules is the best one to choose (some seem not maintained, some might lack API to extract comments,...).

**Comment by hbinkle** [comment]
Is this really a bug?
Regarding to [RFC7149](https://datatracker.ietf.org/doc/html/rfc7159) the JSON file format does not support comments.
All linters I know do not support comments in JSON files.
Especially for WebExtention you don't need comments.
WebExtension supports the `description` beside the action translation which is meant to the the hint for developer and translater. And that `description` is supported/displayed by Weblate.
That's the advantage of the WebExtension format over usual JSON format with key value pairs only.

**Comment by rugk** [comment]
Yes it is a bug,  because [as indicated in the spec (draft) and docs](https://browserext.github.io/browserext/#manifest-json) (see all linked stuff in the initial post) actually **comments are supported**. It is not the usual JSON you refer to, it is "JSON with comments".

> All linters I know do not support comments in JSON files.

Many linters support it. E.g. in VSCode you can change the format to "JSON with comments" and it is supported.

> WebExtension supports the description 

Yeah, that's a good thing, but what I personally use them for is grouping the translations by giving them a header. See https://github.com/Lusito/web-ext-translator/issues/10 for an implementation in a translation tool, which is quite nice and is what I currently use…

**Comment by hbinkle** [comment]
@rugk, I don't know what you expect "https://github.com/browserext/browserext" should be. But that's no official RFC rather than an official RFC draft. It even doesn't have a useful description/technical spec how/which comments are allowed.
If you reference such a page you may also have a look at wikipedia: https://en.wikipedia.org/wiki/JSON which state that JSON is supposed NOT to have comments.

> I personally use them for is grouping
If you abuse the description of the webExtension format for grouping and use it for hints/comments as it's meant to be used for, you should not complain about/request missing comments.

**Comment by rugk** [comment]
> If you abuse the description of the webExtension format for grouping and use it for hints/comments as it's meant to be used for, 

As far as I see there is a misunderstanding: I do use the description _as intended_ for comments/a description. I do use the _comments_ for grouping though.
That's just to clarify/show one use-case these comments have.

> But that's no official RFC rather than an official RFC draft.

Yes, I know it's a draft and all that, but actually, it describes _how it is currently implemented in browsers_, actually. For more information see https://wiki.mozilla.org/WebExtensions/Spec
All major browsers (Chrome, Edge, Opera and Firefox) do accept JSON with comments AFAIK and as said [IDEs also have a way to accept it](https://code.visualstudio.com/Docs/languages/json#_json-with-comments) (if you need look up there any explanation).

I know JSON does not support comments, but WebExtension's derived JSON format (which you may call differently), actually does. As this issue asks about support for "WebExtension JSON" (maybe call it like this), it of course should support this _variant_ of such a markup format.
